### PR TITLE
Delivering all outputs

### PR DESCRIPTION
**Pull request by omokaro-faith**
What does this PR do?

Create output for writing test
Create output for setting up software development environment
Create output for getting peer review
Create output for setting-up linters
Create output for automating tasks
Create output for checking repository health
Create output for delivering the product
Create output for reflecting on the learning experience
Create output for setting up continuous integration
Create output for setting up repository
Create output for solving the problem
Create output for using design pattern
Create output for HTML-CSS-js

Description of Task to be completed?
Provide:
- A screenshot of your code coverage and GPA badges in your repo's README file,
- A link to a successful TravisCI build of your development branch (it should include test runs),
- Live URLs to a review app, your staging & production apps
- A Pull Request showing at least two Peer Code Reviews
- The repository for your proposed project and at least one Pull Request
- A post reflecting on your learning experience while producing these outputs
- Release notes for your final product in your repo's wiki
- A detailed explanation of Design Patterns used while solving-the-problem
- A folder containing the template for the proposed front-end of your project
- A link to a video showing how gulp automates tasks
- A folder containing detailed explanation of Design Patterns used while `solving-the-problem`

How should this be manually tested?
For:
- checking repository health: visit the [output folder](https://github.com/andela/test-simulations-mgmt-repo/tree/mark-aragon-fomokaro/06-building-a-program-start-to-finish/checking-repository-health/Output), it contains the link 
- setting up continuous integration with Travis CI: open the successful Travis [build link](https://travis-ci.org/andela-fomokaro/Inverted-Index-Project)
- setting up software development environments: open the [review](https://fomokaro-invertedindex-st-pr-2.herokuapp.com/), [staging](https://fomokaro-invertedindex-staging.herokuapp.com/) and [production](https://fomokaro-inverted-production.herokuapp.com/) links
- getting peer reviews:
[link to pull request showing peer code reviews on pullrequest-1](https://github.com/andela-fomokaro/Inverted-Index-Project/pull/2)
[link to pull request showing peer code reviews on pullrequest-2](https://github.com/andela-fomokaro/Inverted-Index-Project/pull/1)
- setting up the repository: open the [repository link](https://github.com/andela-fomokaro/Inverted-Index-Project)
- reflecting on the learning experience: read the [blog post](https://medium.com/@faith.omokaro/checkpoints-and-simulations-my-road-to-learning-38c3ccf07c09) on the subject
- delivering your product: open the [release notes](https://docs.google.com/document/d/1kbvLKewvndFRu_7ez4qlVHpyv6Kun1CzDr0X-WlOiuw/edit)
- using design patterns: visit the [output folder](https://github.com/andela/test-simulations-mgmt-repo/tree/mark-aragon-fomokaro/06-building-a-program-start-to-finish/using-design-patterns/Output), it contains the explanation
- writing front end template (html, css, js): visit the [output folder](https://github.com/andela/test-simulations-mgmt-repo/tree/mark-aragon-fomokaro/06-building-a-program-start-to-finish/writing-html-js-css-outputs-(template)/Output), it contains the front-end template

Any background context you want to provide?

NA

What are the relevant pivotal tracker stories?

NA

Screenshots (if appropriate)

NA

Questions:

NA

### CODE REVIEW

File: 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+];
+describe('InvertedIndex : define methods', () => {
+  invertedIndex.createIndex('work.json', workBook);
+  it('Should ensure that invertedIndex methods are defined', () => {
+    expect(invertedIndex.getIndex).toBeDefined();
+    expect(invertedIndex.createIndex).toBeDefined();
+    expect(invertedIndex.tokenize).toBeDefined();
+    expect(Index.validateFile).toBeDefined();
+    expect(invertedIndex.searchIndex).toBeDefined();
+    expect(invertedIndex.flattenSearch).toBeDefined();
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js]: Please use a separate it block for each of these.

File: 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        one: [0],
+        testing: [0],
+        getindex: [0],
+        function: [0],
+        two: [1],
+        coming: [1],
+        soon: [1],
+      });
+    });
+  it('should verify that index has been created', () => {
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js]: Why are you repeating this speecification?

File: 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  it('Should return correct index of the search term in work.json', () => {
+    expect(invertedIndex.searchIndex('soon', 'work.json')).toEqual({
+      soon: [1],
+    });
+  });
+  it('should return true if the constant is a string', () => {
+    const words = 'Alice in Wonderland';
+    expect(Object.keys(invertedIndex.searchIndex(words))).toBeTruthy();
+  });
+  it('Should return correct index of the search term', () => {
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js]: Why are you repeating this specification?

File: 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    expect(invertedIndex.searchIndex('humans', 'test.json')).toEqual({
+      humans: [1],
+    });
+  });
+  it('should return tokenized words for string in variable words', () => {
+    let words = 'It CAN only BE++ $$$$$   GOD';
+    const termsExpected = 'it can only be god';
+    words = invertedIndex.tokenize(words);
+    expect(termsExpected).toEqual(words);
+  });
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js]: Should this it block be part of your searchIndex() Test Suite or tokenize() Test Suite?

File: 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+describe('InvertedIndex : Read file data ', () => {
+  it('should read and return the contents of a file via callback',
+   (read) => {
+     const file = new File('./files/test.json');
+     Index.readFile(file, (e) => {
+       expect(JSON.parse(e.target.result)).toEqual(test);
+       read();
+     });
+   });
+});
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js]: Your test are not properly done. Repetition of test suites and specifications. Please review it

**Review by andela-moe — COMMENTED** [review]
Please address comments.

File: 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js (hunk 3 above)
Comment (omokaro-faith) [reply 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js]: Made changes to these file

File: 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js (hunk 2 above)
Comment (omokaro-faith) [reply 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js]: I have made changes to these file

File: 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js (hunk 4 above)
Comment (omokaro-faith) [reply 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js]: I made changes already

File: 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js (hunk 5 above)
Comment (omokaro-faith) [reply 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js]: I have made changes where it was needed to the file

File: 06-building-a-program-start-to-finish/checking-repository-health/Output/README.md [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#### Output
+- **Using**: `Code Climate` and `Coveralls`
+- **Deliver**: a screenshot of your code coverage and GPA badges in your repo's README file
+- **With**: *independently*
+
+#### Targeted Learning Outcomes
+[link to travis and coverage badge](https://github.com/andela-fomokaro/Inverted-Index-Project/blob/develop/Pictures%20of%20eslint%20error%20violation%20and%20houndci/travis%20and%20coverage%20badges.png)
+
+### Author
+### Omokaro Faith
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/checking-repository-health/Output/README.md:10]: I see your badges. Why is coverage 77%?

File: 06-building-a-program-start-to-finish/automating-tasks/Output/README.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
+#### Output
+- **Using**: Gulp
+- **Deliver**: a screencast showing gulp watching and refreshing your browser when changes are made to your HTML, CSS & JS files
+- **With**: *independently*
+
+#### Targeted Learning Outcomes
+[link to screencast showing gulp watching HTML, CSS & JS files !](https://www.youtube.com/watch?v=QxiJnVfEKkQ)
+
+### Author
+### Omokaro Faith
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/automating-tasks/Output/README.md:10]: Cool

File: 06-building-a-program-start-to-finish/delivering-your-product/Output/README.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
+#### Output
+- **Using**: Google Doc
+- **Deliver**: `Release notes` for your final product in your repo's wiki
+- **With**: *independently*
+
+#### Targeted Learning Outcomes
+[Link to release note!](https://docs.google.com/a/andela.com/document/d/1kbvLKewvndFRu_7ez4qlVHpyv6Kun1CzDr0X-WlOiuw/edit?usp=sharing)
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/delivering-your-product/Output/README.md]: Cool. Please add to your repo's wiki

File: 06-building-a-program-start-to-finish/getting-peer-reviews/Output/README.md [hunk 9]
```diff
@@ -0,0 +1,8 @@
+#### Output
+- **Using**: GitHub
+- **Deliver**: a Pull Request showing at least two Peer Code Reviews
+- **With**: *independently*
+
+#### Targeted Learning Outcomes
+- [link to pull request showing peer code reviews on pullrequest-1](https://github.com/andela-fomokaro/Inverted-Index-Project/pull/2)
+- [link to pull request showing peer code reviews on pullrequest-2](https://github.com/andela-fomokaro/Inverted-Index-Project/pull/1)
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/getting-peer-reviews/Output/README.md:8]: Nice

File: 06-building-a-program-start-to-finish/reflecting-on-the-learning-experience/Output/README.md [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+- **Deliver**: a post reflecting on your learning experience while producing these outputs
+- **With**: *independently*
+
+#### Targeted Learning Outcomes
+#### Output
+
+[Link to Medium Post!](https://medium.com/@faith.omokaro/checkpoints-and-simulations-my-road-to-learning-38c3ccf07c09#.e27qrpifq)
+
+## Author
+### Omokaro Faith
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/reflecting-on-the-learning-experience/Output/README.md:12]: Good read

File: 06-building-a-program-start-to-finish/setting-up-linters/Output/README.md [hunk 11]
```diff
@@ -0,0 +1,8 @@
+#### Output
+- **Using**: Terminal & GitHub
+- **Deliver**: two screenshots of ESLint [Airbnb](https://github.com/airbnb/javascript)  style violations in the terminal and [HoundCI](https://houndci.com/) comment on a PR
+- **With**: *independently*
+
+#### Targeted Learning Outcomes
+ [link to Eslint violation for Houndci](https://github.com/andela-fomokaro/Inverted-Index-Project/blob/features/Pictures%20of%20eslint%20error%20violation%20and%20houndci/Screen%20Shot%202017-03-26%20at%207.43.05%20AM.png) style and
+ [link to Eslint violation for Airbnb](https://github.com/andela-fomokaro/Inverted-Index-Project/blob/features/Pictures%20of%20eslint%20error%20violation%20and%20houndci/Screen%20Shot%202017-03-26%20at%207.20.53%20AM.png) outputs
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/setting-up-linters/Output/README.md:8]: Cool.

File: 06-building-a-program-start-to-finish/setting-up-software-development-environments-(deployment)/Output/README.md [hunk 12]
```diff
@@ -0,0 +1,8 @@
+#### Output
+- **Using**: Heroku
+- **Deliver**: live URLs to a `review` app, your `staging` & `production` apps
+- **With**: *independently*
+
+#### Targeted Learning Outcomes
+live URLs to
+[production](https://fomokaro-inverted-production.herokuapp.com/) [staging](https://fomokaro-invertedindex-staging.herokuapp.com/) and [review](https://fomokaro-invertedindex-st-pr-2.herokuapp.com/) applications.
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/setting-up-software-development-environments-(deployment)/Output/README.md:8]: Live Urls. Nice

File: 06-building-a-program-start-to-finish/setting-up-your-repository/Output/README.md [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+- **Using**: GitHub
+- **Deliver**: the repository for your proposed project and at least one `Pull Request`
+- **With**: *independently*
+
+#### Targeted Learning Outcomes
+Link to Github repo
+https://github.com/andela-fomokaro/Inverted-Index-Project
+
+## Author
+### Omokaro Faith
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/setting-up-your-repository/Output/README.md:11]: Ok

File: 06-building-a-program-start-to-finish/using-design-patterns/Output/README.md [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+MVC is an achitectural pattern that is used to separate an application's concerns. That is, application code is broken into three layers:
+
+Model - represents an object carrying data. e.g. InvertedIndex Class
+View - represents the front end view of the data that model contains. e.g. index.html
+Controller - acts on both model and view. It controls the data flow into model object and updates the view whenever data changes. e.g. app.js
+Hence as explained above, in the case of the Inverted Index App, the model is the InvertedIndex Class, the view is the front end HTML and CSS, while the controller is the angular app.
+
+
+## Author
+### Omokaro Faith
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/using-design-patterns/Output/README.md:85]: Ok

File: 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js [hunk 15]
```diff
@@ -0,0 +1,127 @@
+const Index = require('../Model/inverted-index.js');
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/writing-tests/Output/spec/index-spec.js]: Please correct the name convention for your test file

File: 06-building-a-program-start-to-finish/solving-the-problem/Output/inverted-index.js [hunk 16]
```diff
@@ -0,0 +1,171 @@
+const fileApi = require('file-api');
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/solving-the-problem/Output/inverted-index.js]: Is this (inverted-index.js) the proper naming convention for this file

**Review by andela-moe — COMMENTED** [review]
Please review comments

File: 06-building-a-program-start-to-finish/solving-the-problem/Output/inverted-index.js [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+     *
+     * @method searchIndex
+     * @param {String} searchTerms
+     * @param {String} fileName
+     *
+     * @returns {Object} searchResult
+     */
+  searchIndex(searchTerms, fileName) {
+    const searchResult = {};
+    let tempObject = {};
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/solving-the-problem/Output/inverted-index.js]: Please use a more appropriate name for this variable

File: 06-building-a-program-start-to-finish/solving-the-problem/Output/inverted-index.js [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+     * @method tokenize
+     * @param {string} words
+     * @return {String} this.token
+     */
+  tokenize(words) {
+    this.token = words.toLowerCase()
+    .trim()
+    .match(/\w+/g);
+    return this.token;
+  }
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/solving-the-problem/Output/inverted-index.js]: I your tokenize() method taking care of special characters?

File: 06-building-a-program-start-to-finish/checking-repository-health/Output/README.md (hunk 6 above)
Comment (omokaro-faith) [reply 06-building-a-program-start-to-finish/checking-repository-health/Output/README.md:10]: The badge refused to change

File: 06-building-a-program-start-to-finish/delivering-your-product/Output/README.md (hunk 8 above)
Comment (omokaro-faith) [reply 06-building-a-program-start-to-finish/delivering-your-product/Output/README.md]: ok

File: 06-building-a-program-start-to-finish/solving-the-problem/Output/inverted-index.js (hunk 16 above)
Comment (omokaro-faith) [reply 06-building-a-program-start-to-finish/solving-the-problem/Output/inverted-index.js]: I have made changes to the files

File: 06-building-a-program-start-to-finish/solving-the-problem/Output/inverted-index.js (hunk 18 above)
Comment (omokaro-faith) [reply 06-building-a-program-start-to-finish/solving-the-problem/Output/inverted-index.js]: I have fixed the issue

File: 06-building-a-program-start-to-finish/solving-the-problem/Output/inverted-index.js (hunk 17 above)
Comment (omokaro-faith) [reply 06-building-a-program-start-to-finish/solving-the-problem/Output/inverted-index.js]: I have fixed the issue

File: 06-building-a-program-start-to-finish/checking-repository-health/Output/README.md (hunk 6 above)
Comment (andela-moe) [reply 06-building-a-program-start-to-finish/checking-repository-health/Output/README.md:10]: Please fix this coverage badge before your defense - Still displays 86%.

File: 06-building-a-program-start-to-finish/writing-html-js-css-outputs-(template)/Output/index.html [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+      </div>
+   </body>
+   <script type="text/javascript" src="https://code.jquery.com/jquery-2.1.1.min.js"></script>
+   <script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.6.1/angular.js"></script>
+   <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.1.1/jquery.min.js"></script>
+   <script src="https://cdnjs.cloudflare.com/ajax/libs/toastr.js/latest/toastr.min.js"></script>
+   <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
+   <script src="./Model/inverted-index.js"></script>
+   <script src="./Model/app.js"></script>
+</html>
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/writing-html-js-css-outputs-(template)/Output/index.html]: Have you been able to reorder the layout of this page.

File: 06-building-a-program-start-to-finish/writing-tests/Output/spec/invertedIndex.spec.js [hunk 20]
```diff
@@ -0,0 +1,130 @@
+const InvertedIndex = require('../Model/invertedIndex.js');
+const test = require('../files/test.json');
+const work = require('../files/work.json');
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/writing-tests/Output/spec/invertedIndex.spec.js]: Why declare a variable you are using in this file?

**Review by andela-moe — DISMISSED** [review]
Please review comments

File: 06-building-a-program-start-to-finish/writing-html-js-css-outputs-(template)/Output/index.html (hunk 19 above)
Comment (omokaro-faith) [reply 06-building-a-program-start-to-finish/writing-html-js-css-outputs-(template)/Output/index.html]: Yes

File: 06-building-a-program-start-to-finish/writing-tests/Output/spec/invertedIndex.spec.js (hunk 20 above)
Comment (omokaro-faith) [reply 06-building-a-program-start-to-finish/writing-tests/Output/spec/invertedIndex.spec.js]: I have made changes to file

File: 06-building-a-program-start-to-finish/solving-the-problem/Output/invertedIndex.js [hunk 21]
```diff
@@ -0,0 +1,308 @@
...
+        }
+      });
+      searchResult[filename] = temporaryObject;
+    });
+    return searchResult;
+  }
+}
+
+module.exports = InvertedIndex;
+const fileApi = require('file-api');
```
Comment (andela-moe) [inline 06-building-a-program-start-to-finish/solving-the-problem/Output/invertedIndex.js]: Why do you have this replicated in this file?

**Review by andela-moe — COMMENTED** [review]
Good work. Please remove the duplicate on your invertedIndex.js file.

File: 06-building-a-program-start-to-finish/solving-the-problem/Output/invertedIndex.js (hunk 21 above)
Comment (omokaro-faith) [reply 06-building-a-program-start-to-finish/solving-the-problem/Output/invertedIndex.js]: That was a mistake modifed the file already

**Comment by andela-moe** [comment]
All Good.

**Pull request merged**